### PR TITLE
Pre commit hook

### PR DESCRIPTION
**Pull request by casperdcl**
add pre-commit hook (https://pre-commit.com/)

allows (in any repo):

```bash
$ pip install pre-commit
$ echo '\
```
```yml
repos:
-   repo: https://github.com/mvdan/sh
    rev: master
    hooks:
    -   id: shfmt
```
```bash
' > .pre-commit-config.yaml
$ pre-commit install
```

### CODE REVIEW

File: .pre-commit-hooks.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  language: golang
+  types: [shell]
+  description: 'Formats shell programs'
+  args:
+  - -l  # list files whose formatting differs from shfmt's
+  - -d  # error with a diff when the formatting differs
+  - -i  # indentation
+  - '2' # 2 spaces
+  - -ci  # switch cases will be indented
+  - -w  # write result to file instead of stdout
```
Comment (casperdcl) [inline .pre-commit-hooks.yaml]: note that these are defaults which can be overridden in `.pre-commit-config.yaml`

**Comment by mvdan** [comment]
This was brought up before; see https://github.com/mvdan/sh/pull/465. Please read that thread to get up to speed.

I'm still not convinced that this is necessary. Running the tool is already very simple, via either a binary or a Docker image. It also seems to me like if someone really wants to explicitly support third party tools or software, that could be done in separate git repos. For example, people have written vscode and vim plugins, but they don't live here because I don't maintain them.

Please understand that I appreciate the contribution - I just don't quite get what a small yaml file at the root of the repository is really adding here.

**Comment by casperdcl** [comment]
I don't have a problem leaving it in my fork; just think it would be convenient in case you want to merge it.

pre-commit isn't about making it easier to install, though. It's about, well, pre-commit hooks. Are you against the concept of pre-commit hooks?

Note that related tools (https://github.com/psf/black, https://github.com/pycqa/flake8) all provide them.

**Comment by mvdan** [comment]
I'm not against the concept of pre-commit hooks :) But I am against the concept of having to add a yaml file that doesn't really do anything beyond saying "hey, I support this pre-commit framework!". The tiny amount of data it's actually defining could very well be inferred, or declared by the actual user of pre-commit.

> Note that related tools (https://github.com/psf/black, https://github.com/pycqa/flake8) all provide them.

One of those tools has a dozen "dotfiles" in the root of the repository - I'm not sure that's a good example to strive towards :)

**Comment by sharm294** [comment]
> I'm not against the concept of pre-commit hooks :) But I am against the concept of having to add a yaml file that doesn't really do anything beyond saying "hey, I support this pre-commit framework!". The tiny amount of data it's actually defining could very well be inferred, or declared by the actual user of pre-commit.

This yaml file doesn't just say that you support the pre-commit framework. The .pre-commit-hooks.yaml is needed at the root of repositories that can [serve as hooks](https://pre-commit.com/#creating-new-hooks). It tells pre-commit how to install and call the tool. The user can't infer this or declare it.

> I'm still not convinced that this is necessary. Running the tool is already very simple, via either a binary or a Docker image. It also seems to me like if someone really wants to explicitly support third party tools or software, that could be done in separate git repos. For example, people have written vscode and vim plugins, but they don't live here because I don't maintain them.

Using shfmt in pre-commit right now requires that shfmt exists as an executable on the PATH, so through the binary like you mentioned (I'm excluding running it from forked repos like casperdcl's). I dislike this approach for pre-commit because it forces all contributors of a repo that uses pre-commit and calls shfmt to have the tool installed. With this yaml, pre-commit handles installing shfmt itself in a private environment and the dependency on contributors becomes that they just need Go installed. Or if the yaml said that the language was docker, then the dependency becomes that they need Docker installed and pre-commit takes care of the rest.

And yes, you're right that this can exist in a forked repo, but it's nice if it exists in the official up-to-date one :)

**Comment by casperdcl** [comment]
I think the issue is `pre-commit` itself should ideally put the effort into maintaining hooks for all tools that could support them. However that's too much work for a FOSS effort... so they've understandably moved to asking maintainers of tools to maintain their own hooks. No, it's not ideal either and I also disdain root clutter but such is life.

**Comment by mvdan** [comment]
> The .pre-commit-hooks.yaml is needed at the root of repositories that can [serve as hooks](https://pre-commit.com/#creating-new-hooks). It tells pre-commit how to install and call the tool. The user can't infer this or declare it.

I'd personally say this is a design flaw. This yaml file only really contains three relevant bits of info - the name of the tool, and the language it's written in. In other words, running the tool is generally as easy as `go run mvdan.cc/sh/v3/cmd/shfmt`. There's no reason why I should be declaring this tiny amount of pretty obvious information.

> With this yaml, pre-commit handles installing shfmt itself in a private environment and the dependency on contributors becomes that they just need Go installed.

I agree that a tooling framwork can be nice. I disagree that it's my job to bend over backwards to add tiny yaml files to make each framework happy, though :)

> However that's too much work for a FOSS effort... so they've understandably moved to asking maintainers of tools to maintain their own hooks.

I get that, but I also have limited time, and I want to minimise the amount of environments and edge cases I need to keep in mind when I maintain the project in the long run. And, as said above, I still don't think this yaml file is adding any useful information.

One can imagine a pre-commit design where one could do:

```
repos:
-   repo: https://github.com/mvdan/sh
    rev: v3.0.2
    hooks:
    -   id: shfmt
        language: golang
```

That's one more line than the current suggestion, and doesn't require upstream to maintain a yaml file. I get that pre-commit doesn't work this way, and I get that they want me to add a yaml file, but I just disagree, sorry.

**Comment by scop** [comment]
FYI, with pre-commit 2.4.0 which was just released, it's now possible to use shfmt with a `local` repo config, for example:
```yaml
  - repo: local
    hooks:
      - id: shfmt
        name: shfmt
        language: golang
        additional_dependencies: [mvdan.cc/sh/v3/cmd/shfmt@v3.1.1]
        entry: shfmt
        args: [-w, -s]
        types: [shell]
```

**Comment by mvdan** [comment]
@scop thanks! want to update your PR to add this to the README? It could be a short section under the Docker one.

**Comment by scop** [comment]
Ah, I had already commented there, but can do tomorrowish.

**Comment by kenorb** [comment]
This is still not implemented. I've tested with:

```
  - repo: https://github.com/mvdan/sh
    rev: master
    hooks:
      - id: shfmt
        name: shfmt
        minimum_pre_commit_version: 2.4.0
        language: golang
        additional_dependencies: [mvdan.cc/sh/v3/cmd/shfmt@v3.1.1]
        entry: shfmt
        args: [-w]
        types: [shell]
```

I've used this repo instead: https://github.com/jumanjihouse/pre-commit-hooks

```
  - repo: https://github.com/jumanjihouse/pre-commit-hooks
    rev: master
    hooks:
      - id: shfmt
```

**Comment by scop** [comment]
See https://github.com/mvdan/sh/blob/master/README.md#pre-commit (TL;DR you need to use `local` as repo, and remove `rev`).

The advantage of this hook over the jumanjihouse one is automatic installation and versioning capability, for their version one needs to take care that the desired shfmt is installed manually/separately.

**Pull request closed without merging**